### PR TITLE
Work with PriceSet throughout `checkout/service.py`

### PR DESCRIPTION
**Pull request by pieterbeulque**
Refactors checkout so seat handling no longer leans on the single `product_price` foreign key. Use the `PriceSet` directly and run all amount math through one pure helper `calculate_upfront_amount`.

This is a pure refactor with no behavior change. For seat-only products the FK is the seat price today, so looking it up from the set returns the same object and the sums collapse to today's single-price math. The existing seat tests pass without any changes.

Rationale: it's the groundwork for combining seat-based prices with a fixed base fee. Once that exists, `calculate_upfront_amount` can just sum both instead of having to pick one.

As a QOL improvement, also changed some custom-price checks to pull the type-check on `product_price` completely out of the `update` logic to streamline things.

---
## Summary by cubic
Refactor checkout to use `PriceSet` end-to-end and compute upfront totals with `calculate_upfront_amount`. Fix PWYW handling to honor `0` presets and `?amount=0` link prefills; behavior is otherwise unchanged and this prepares combining seat prices with a base fee.

- **Refactors**
  - Use `PriceSet` across create, update, and checkout‑link; explicit price selection becomes a single‑price set.
  - Compute upfront totals via `calculate_upfront_amount` from the set’s static prices, plus optional `seats` and custom amount.
  - Resolve seat/custom prices with `PriceSet` helpers; validate limits; default seats to the minimum where needed.

- **Tests**
  - Add unit tests for `PriceSet` helpers and `calculate_upfront_amount`, including honoring `preset_amount=0` and summing fixed + seat prices.
  - Add checkout‑link regression test for `?amount=0` PWYW prefills and a test for minimum‑seat defaults on seat‑based pricing.

<sup>Written for commit d6d4ce42354b70c57915fbaa66c5bb0307a6ed4b. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/polarsource/polar/pull/12160?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: server/polar/product/price_set.py [hunk 1]
```diff
@@ -92,9 +99,59 @@ def get_default_price(self) -> ProductPrice:
...
+    """
+    amount = 0
+    for price in prices:
+        if is_fixed_price(price):
+            amount += price.price_amount
+        elif is_custom_price(price):
+            if custom_amount is not None:
+                amount += custom_amount
+            else:
+                amount += price.preset_amount or price.minimum_amount
```
[the remark this answers is not in the document]
Comment (frankie567) [reply server/polar/product/price_set.py]: It's right!

File: server/polar/checkout/service.py [hunk 2]
```diff
@@ -713,28 +710,32 @@ async def checkout_link_create(
...
                     pass
 
-            amount = valid_query_amount or price.preset_amount or price.minimum_amount
-        elif is_seat_price(price):
-            # Default to minimum seats for checkout links with seat-based pricing
-            seats = price.get_minimum_seats()
-            amount = price.calculate_amount(seats)
+        amount = calculate_upfront_amount(
+            currency_prices.get_static_prices(),
+            custom_amount=custom_amount or None,
```
[the remark this answers is not in the document]
Comment (frankie567) [reply server/polar/checkout/service.py]: Yes!

**Review by frankie567 — COMMENTED** [review]
Yes, that really makes sense. Just need to fix the comments from the AI, which are correct.

**Comment by pieterbeulque** [comment]
@frankie567 - I fixed them in a separate commit now. It looks like both were pre-existing bugs

**Review by frankie567 — APPROVED** [review]
